pkg/util/parse/grpc.go

```diff
@@ -0,0 +1,346 @@
+				visitor = *v
+				break
+			}
+		}
+
+		protocolName, err := getProtocolNameFromGrpc(device)
+		if err != nil {
+			klog.Errorf("fail to get protocol name from grpc for device %s with err: %+v", device.Name, err)
+			return nil
+		}
+
+		var visitorConfig []byte
+		switch protocolName {
+		case controller.Modbus:
+			visitorConfig, err = json.Marshal(visitor.Modbus)
+			if err != nil {
+				return nil
+			}
+		case controller.OPCUA:
+			visitorConfig, err = json.Marshal(visitor.Opcua)
+			if err != nil {
+				return nil
+			}
+		case controller.Bluetooth:
+			visitorConfig, err = json.Marshal(visitor.Bluetooth)
+			if err != nil {
+				return nil
+			}
+		case controller.CustomizedProtocol:
+			visitorConfig, err = json.Marshal(visitor.CustomizedProtocol)
+			if err != nil {
+				return nil
+			}
+		}
+
+		cur := common.Twin{
+			PropertyName: twin.PropertyName,
+			PVisitor: &common.PropertyVisitor{
+				Name:         twin.PropertyName,
+				PropertyName: twin.PropertyName,
+				ModelName:    device.Spec.DeviceModelReference,
+				CollectCycle: visitor.CollectCycle,
+				ReportCycle:  visitor.ReportCycle,
+				PProperty: common.Property{
+					Name:         twin.PropertyName,
+					DataType:     "",
+					Description:  "",
+					AccessMode:   "",
+					DefaultValue: nil,
+					Minimum:      0,
+					Maximum:      0,
+					Unit:         "",
+				},
+				Protocol:      protocolName,
+				VisitorConfig: visitorConfig,
+			},
+			Desired: common.DesiredData{
+				Value: twin.Desired.Value,
+				Metadatas: common.Metadata{
+					Timestamp: twin.Desired.Metadata["timestamp"],
+					Type:      twin.Desired.Metadata["type"],
+				},
+			},
+			Reported: common.ReportedData{
+				Value: twin.Reported.Value,
+				Metadatas: common.Metadata{
+					Timestamp: twin.Desired.Metadata["timestamp"],
+					Type:      twin.Desired.Metadata["type"],
+				},
+			},
+		}
+		res = append(res, cur)
+	}
+	return res
+}
+
+func buildDataFromGrpc(device *dmiapi.Device) common.Data {
+	res := common.Data{}
+	if len(device.Spec.PropertyVisitors) > 0 {
+		res.Properties = make([]common.DataProperty, 0, len(device.Spec.PropertyVisitors))
+		for _, property := range device.Spec.PropertyVisitors {
+			cur := common.DataProperty{
+				Metadatas:    common.DataMetadata{},
+				PropertyName: property.PropertyName,
+				PVisitor:     nil,
+			}
+			if property.CustomizedValues != nil && property.CustomizedValues.Data != nil {
+				timestamp, ok := property.CustomizedValues.Data["timestamp"]
+				if ok {
+					t, _ := strconv.ParseInt(string(timestamp.GetValue()), 10, 64)
+					cur.Metadatas.Timestamp = t
+				}
+				tpe, ok := property.CustomizedValues.Data["type"]
+				if ok {
+					cur.Metadatas.Type = string(tpe.GetValue())
+				}
+				res.Properties = append(res.Properties, cur)
+			}
+		}
+	}
+	return res
+}
+
+func buildPropertyVisitorsFromGrpc(device *dmiapi.Device) []common.PropertyVisitor {
+	klog.Infof("##########################")
```

Comment:
these log in this file is for test. we can remove them.

Reply:
done